sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -0,0 +1,1704 @@
+ * with extra locking is the `SpillableHostStore`, to maintain a `totalSize` number that is
+ * used to figure out cheaply when it is full.
+ *
+ * All handles, except for disk handles, hold a reference to an object in their respective store:
+ * `SpillableDeviceBufferHandle` has a `dev` reference that holds a `DeviceMemoryBuffer`, and a
+ * `host` reference to `SpillableHostBufferHandle` that is only set if spilled. Disk handles are
+ * different because they don't spill, as disk is considered the final store. When a user calls
+ * `materialize` on a handle, the handle must guarantee that it can satisfy that, even if the caller
+ * should wait until a spill happens. This is currently implemented using the handle lock.
+ *
+ * Note that we hold the handle lock while we are spilling (performing IO). That means that no other
+ * consumer can access this spillable device handle while it is being spilled, including a second
+ * thread that is trying to spill and is generating a spill plan, as the handle lock is likely held
+ * up with IO. We will relax this likely in follow on work.
+ *
+ * We never hold a store-wide coarse grain lock in the stores when we do IO.
+ */
+
+/**
+ * Common interface for all handles in the spill framework.
+ */
+trait StoreHandle extends AutoCloseable {
+  /**
+   * Approximate size of this handle, used in three scenarios:
+   * - Used by callers when accumulating up to a batch size for size goals.
+   * - Used from the host store to figure out how much host memory total it is tracking.
+   * - If approxSizeInBytes  is 0, the object is tracked by the stores so it can be
+   *   removed on shutdown, or by handle.close, but 0-byte handles are not spillable.
+   */
+  val approxSizeInBytes: Long
+}
+
+trait SpillableHandle extends StoreHandle {
+  /**
+   * Method called to spill this handle. It can be triggered from the spill store,
+   * or directly against the handle.
+   *
+   * This will not free the spilled data. If you would like to free the spill
+   * call `releaseSpilled`
+   *
+   * @note The size returned from this method is only used by the spill framework
+   *       to track the approximate size. It should just return `approxSizeInBytes`, as
+   *       that's the size that it used when it first started tracking the object.
+   * @return approxSizeInBytes if spilled, 0 for any other reason (not spillable, closed)
+   */
+  def spill(): Long
+
+  /**
+   * Part two of the two-stage process for spilling. We call `releaseSpilled` after
+   * a handle has spilled, and after a device synchronize. This prevents a race
+   * between threads working on cuDF kernels, that did not synchronize while holding the
+   * materialized handle's refCount, and the spiller thread (the spiller thread cannot
+   * free a device buffer that the worker thread isn't done with).
+   * See https://github.com/NVIDIA/spark-rapids/issues/8610 for more info.
+   */
+  def releaseSpilled(): Unit
```

Comment:
This interface is specific to device handles, and it's suboptimal to force non-device handles to separate spilling from releasing memory, e.g.: multi-part spillables that could free parts as they are spilled.

Reply:
good point, let me propose a different API for the spill stores that should be more flexible.